cloud_env.go

```diff
@@ -0,0 +1,325 @@
+/*
+ * © Copyright IBM Corp. 2018
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package IBMCloudEnv
+
+import ("os"
+		"fmt"
+		"strings"
+  	"github.com/tidwall/gjson"
+  	"io/ioutil"
+ 		"github.com/oliveagle/jsonpath"
+ 		"encoding/json"
+ 		log "github.com/sirupsen/logrus"
+)
+
+const PREFIX_PATTERN_CF = "cloudfoundry"
+const PREFIX_PATTERN_ENV = "env"
+const PREFIX_PATTERN_FILE = "file"
+const PREFIX_PATTERN_USER = "user-provided"
+
+var loadedMappings = make (map [string]interface{}) 
+
+func check(e error) {
+    if e != nil {
+    	log.Error(e)
+    }
+}
+
+func strip_leading_slash(path string) string {
+	if (len(path) > 0) && (path[0] == '/' || path[0] == '\\'){
+		return path[1:]
+	}    
+    return path
+}
+    
+func Initialize(mappingsFilePath string) string {
+	json, err := ioutil.ReadFile(mappingsFilePath)
+  if err != nil {
+  	log.Error(err)
+  }
+	dir, err := os.Getwd()
+	if err != nil {
+  	log.Error(err)
+	}
+  mappingsFilePath =  dir + mappingsFilePath;
+  result := gjson.Parse(string(json))
+  version := result.Get("version").Int()
+  result.ForEach(func (key, value gjson.Result) bool{
+  	if !result.Get("version").Exists() {
+	  	processMapping(key.String(),value) 
+	  } else if version == 1 {
+	  	processMapping(key.String(),value) 
+	  } else if version == 2{
+	  	processMappingV2(key.String(),value) 
+	  }
+  	return true 
+  })
+	return mappingsFilePath
+}
+
+func processMapping(mappingName string, config gjson.Result) bool {
+	searchPatterns := config.Get("searchPatterns")
+	if !searchPatterns.Exists() || len (searchPatterns.Array()) == 0{
+		log.Warnln("No searchPatterns found for mapping", mappingName)
+		return false
+	}
+	searchPatterns.ForEach (func (_, searchPattern gjson.Result) bool {
+		value, OK := processSearchPattern(mappingName, searchPattern.String())
+		if OK {
+			loadedMappings[mappingName] = value
+			return false
+		} else {
+			return true
+		}
+
+	})
+
+	return true
+}
+
+func processMappingV2(mappingName string, config gjson.Result){
+	config.ForEach(func (key, value gjson.Result) bool {
```

Comment:
Is this syntax right? I'm not familiar, but it looks weird that the bool return type isn't on line 94.

Reply:
This is actually fine. How that ForEach function works is that it takes a lambda function with a bool return type, so those return bools are for those ForEach lambda functions, not the processMappingV2